### ISSUE TITLE
XML misc enum issues

### ISSUE THREAD

**Issue by mdsitton** [KhronosGroup/Vulkan-Docs#6]
The following section of vk.xml has some issues if one was wanting to use the file to generate a binding for non C/C++ languages. Because the language target may not support it properly. Also technically it seems the LL extended literals came with C99 and C++11 so this *could* be an issue even with the target language if you have some odd compiler maybe? idk.

Plus its just weird that these values would need to be computed in C99/C++11 to get the proper value for them....
```
<enums name="API Constants" comment="Misc. hardcoded constants - not an enumerated type">
        <!-- This is part of the header boilerplate -->
    <enum value="256"   name="VK_MAX_PHYSICAL_DEVICE_NAME_SIZE"/>
    <enum value="16"    name="VK_UUID_SIZE"/>
    <enum value="256"   name="VK_MAX_EXTENSION_NAME_SIZE"/>
    <enum value="256"   name="VK_MAX_DESCRIPTION_SIZE"/>
    <enum value="32"    name="VK_MAX_MEMORY_TYPES"/>
    <enum value="16"    name="VK_MAX_MEMORY_HEAPS"/> <!-- The maximum number of unique memory heaps, each of which supporting 1 or more memory types. -->
    <enum value="1000.0f" name="VK_LOD_CLAMP_NONE"/>
    <enum value="(~0U)" name="VK_REMAINING_MIP_LEVELS"/>
    <enum value="(~0U)" name="VK_REMAINING_ARRAY_LAYERS"/>
    <enum value="(~0ULL)" name="VK_WHOLE_SIZE"/>
    <enum value="(~0U)" name="VK_ATTACHMENT_UNUSED"/>
    <enum value="1"     name="VK_TRUE"/>
    <enum value="0"     name="VK_FALSE"/>
    <enum value="(~0U)" name="VK_QUEUE_FAMILY_IGNORED"/>
    <enum value="(~0U)" name="VK_SUBPASS_EXTERNAL"/>
</enums>
```

**Comment by oddhack** [comment]
We did consciously restrict the Vulkan header to modern C/C++ compilers, after considerable polling of IHVs and developers supporting different platforms and compilers, The contortions we go through in the GL XML/headers to deal with 64-bit types were a strong discincentive.

You're right that there are C-specific constructs in gl.xml, and not just here - the param and member tags for commands and structures contain (annotated with type information) C declarations, and some of the type tags contain C #defines . This was a design choice. The original SGI gl.spec file used a notation that was language independent, which led to pretty horrid contortions both converting C APIs into .spec format, and going back the other way, as well as an almost unreadable syntax. When I did the GL XML schema, I chose to cut out all the conversions and just use C declarations, with the expectation that people targeting other languages would have to write a parser. Hopefully only once, and someone will contribute it back to the project.

So I'd be happy to accomodate other languages, but I don't want to revisit the C-centric syntax without really strong motivation, when I think their needs can be accomodated in the parser scripts. Some proofs-of-concept and examples of what you want to do may help drive further discussion on this.

**Comment by mdsitton** [comment]
So one question about intent with those? Are those values intended to compiler specific? Or do they have a specific set value that all implementations of the spec assume.

**Comment by oddhack** [comment]
The constants are not intended to be compiler-specific. In particular the (~0U) constants are a way of getting either the maximum possible unsigned value, or a bitmask with all bits set. In cases where the spec allows variability, there are queries to tell you how many of a certain type of thing there are, and variants of those queries to return up to a specified number of that thing. Some of the other constants are used as the size of arrays in Vk*CreateInfo structures, which must have the same layout between different drivers and compilers to honor the ABI.

**Comment by mdsitton** [comment]
I understand what it does, i have an alternate idea for defining these and wanted to be certain that it would be a good fit.  So perhaps  an attribute that would define how many bits are set for that enum could work?

**Comment by oddhack** [comment]
That would probably work. I am intending to stick a type attribute on the constant definitions at some point, which will remove some ambiguity (and the need to stick XML-escaped ASCII quotes around string constants).

**Comment by davexunit** [comment]
The problem with embedding C into the spec is that now the spec is *not* really XML.  I can't use an XML parser and get a complete tree.  There's still other languages to parse, one of which being a subset of C.  There's also comma delimited lists inside attributes where a sequence of XML nodes would have been more useful.  It's sad to see that the response to these issues seems to be "won't fix".  If someone produced a patch to fix this, would you accept it?

**Comment by vchuravy** [comment]
@oddhack I am currently writing a generator for vk.xml to Julia and even though it took some time I was certainly doable to output the correct bindings.

How do you feel of adding additional generators back to the main branch/repository? Currently my generator lives at https://github.com/JuliaGPU/Vulkan-Docs/tree/vc/julia

**Comment by oddhack** [comment]
@vchuravy I will take a look at your repo soon - sounds like a great mod. I will probably want to partition the contributed additional generators in some fashion (put them in their own files/subdirectory, maybe). Perhaps we can iterate on the scripts to make a cleaner plugin style for the generators?

Let's split this off to a separate Github issue, where I'd be interested to hear what sort of issues you ran into doing this. TBH I know nothing about Julia and hope to learn something there as well.

**Comment by mdsitton** [comment]
So thought I would ask about this, are there any plans to resolve this issue in the foreseeable future?

**Comment by oddhack** [comment]
Forseeable yes, immediate no.
